### PR TITLE
Parse uncertainties

### LINKED ISSUE #1611: Pint + uncertainties in 2022
Pint, Pandas, and Pint-Pandas have all made such great strides in the past year.  With just a little more work, I believe we can support uncertainties in Pint in a clean way using PintPandas (see https://github.com/hgrecco/pint-pandas/issues/139).  Alas, Pint trips itself when passing notes to itself because the formatting of uncertainties is not actually Python.  When the Python tokenizer sees something like `4.4+/-0.9 meter` the parser complains `missing unary operator "/"`

What unary operator, you might ask?  The `/` between the `+` and the `-`.

Here's the testcase:

```
import pytest

from pint import DimensionalityError
from pint.testsuite import QuantityTestCase, helpers

from pint.facets.plain.quantity import PlainQuantity as Quantity
from pint.facets.measurement.objects import Measurement

from pint.compat import HAS_UNCERTAINTIES
assert(HAS_UNCERTAINTIES)
from uncertainties import ufloat, UFloat
from uncertainties import unumpy as unp

Q_ = Quantity
M_ = Measurement

@helpers.requires_uncertainties()
class TestUncertainties(QuantityTestCase):
    def test_uncertainties_fmt(self):
        from numpy.testing import assert_almost_equal
	from uncertainties import ufloat

        u1 = ufloat(1.2, 0.34)
        u2 = ufloat(5.6, 0.78)
        q1_u = module_registry.Quantity(u2 - u1, "m")
        q1_str = str(q1_u)
        q1_str = "{:.4uS}".format(q1_u)
        q1_m = q1_u.magnitude
        q2_u = module_registry.Quantity(q1_str)
        # Not equal because the uncertainties are differently random!                                                                                                                                                                                                                   
        assert q1_u != q2_u
        q2_m = q2_u.magnitude

        assert_almost_equal(q2_m.nominal_value, q1_m.nominal_value, decimal=9)
        assert_almost_equal(q2_m.std_dev, q1_m.std_dev, decimal=4)

	q3_str = "12.34(5678)e-066 m"
        q3_u = module_registry.Quantity(q3_str)
	q3_m = q3_u.magnitude
        assert q3_m < 1

```

### PR DESCRIPTION
**Pull request by MichaelTiemannOSC**
- [x] Closes #1611, #1614
- [x] Executed ``pre-commit run --all-files`` with no errors
- [ ] The change is fully covered by automated unit tests
- [ ] Documented in docs/ as appropriate
- [x] Added an entry to the CHANGES file

I'd like to discuss and review before committing to "completely cover" all automated test units.  I have contributed two test cases that clearly show the filed issues working with the changes.  I have run the full test suite, and the one unexpected failure (#783) I cannot explain.

I'm also not quite sure how to approach the docs question.  Pint documentation tends to be more on the minimal side, and I don't want to call unnecessary attention to things that should "just work".

### CODE REVIEW

**Comment by MichaelTiemannOSC** [comment]
Related: https://github.com/hgrecco/pint-pandas/pull/140

**Comment by hgrecco** [comment]
First, I completely agree with the need of parsing measurements in Pint. I have been looking for this for a while. So thanks, and awesome work.

Having said this, I am not sure about the implementation. Why parsing `8**4` is structurally different from parsing `8+/-4`?

In both cases, you have a couple of numbers “connected” by a multichar operator. I guess that the difference is that `**` is known by the python tokenizer, while `+/-` is not.

Take a look at this:
```python
>>> from pint.compat import tokenizer
>>> for p in tokenizer("8**4"):
...  print(p)
...
TokenInfo(type=2 (NUMBER), string='8', start=(1, 0), end=(1, 1), line='8**4')
TokenInfo(type=54 (OP), string='**', start=(1, 1), end=(1, 3), line='8**4')
TokenInfo(type=2 (NUMBER), string='4', start=(1, 3), end=(1, 4), line='8**4')
TokenInfo(type=4 (NEWLINE), string='', start=(1, 4), end=(1, 5), line='')
TokenInfo(type=0 (ENDMARKER), string='', start=(2, 0), end=(2, 0), line='')
```

So you changed the tokenizer, but in that process it became more difficult to understand, maintain and debug. I wonder if a more sensible approach would not be to add another layer on top of the tokenizer that scans for `+/-`.
 
To understand my proposal, let me show you the output of the current tokenizer:
```python
>>> for p in tokenizer("8 + / - 4"):
...  print(p)
...
TokenInfo(type=2 (NUMBER), string='8', start=(1, 0), end=(1, 1), line='8 + / - 4')
TokenInfo(type=54 (OP), string='+', start=(1, 2), end=(1, 3), line='8 + / - 4')
TokenInfo(type=54 (OP), string='/', start=(1, 4), end=(1, 5), line='8 + / - 4')
TokenInfo(type=54 (OP), string='-', start=(1, 6), end=(1, 7), line='8 + / - 4')
TokenInfo(type=2 (NUMBER), string='4', start=(1, 8), end=(1, 9), line='8 + / - 4')
TokenInfo(type=4 (NEWLINE), string='', start=(1, 9), end=(1, 10), line='')
TokenInfo(type=0 (ENDMARKER), string='', start=(2, 0), end=(2, 0), line='')
```

If we can create function that transform the previous list into:
```python
>>> for p in tokenizer("8 + / - 4"):
...  print(p)
...
TokenInfo(type=2 (NUMBER), string='8', start=(1, 0), end=(1, 1), line='8 + / - 4')
TokenInfo(type=<NEW NUMBER> (OP), string='+ / -', start=(1, 2), end=(1, 7), line='8 + / - 4')
TokenInfo(type=2 (NUMBER), string='4', start=(1, 8), end=(1, 9), line='8 + / - 4')
TokenInfo(type=4 (NEWLINE), string='', start=(1, 9), end=(1, 10), line='')
TokenInfo(type=0 (ENDMARKER), string='', start=(2, 0), end=(2, 0), line='')
```
(this could also handle the parenthesis)

Then the new tokenizer will just be:
```python
def tokenizer(s):
    return uncertainty_layer(old_tokenizer(s))
```

(the names are crap, but just as an example) So it is easy to debug and turn on/off if necessary.

If you go this way, I think it will be wise to move the `tokenizer` function out of the compat module and into pint_eval.

**Comment by hgrecco** [comment]
I also think we need to clearly define which uncertainties format we are going to support (and which we are not).

**Comment by MichaelTiemannOSC** [comment]
Thank you so much for giving it a look.

On the choice of creating a tighter binding than **, consider `x ** 2+/-0 ** (1/2)'.  If +/- binds the same as **, it will fail when there's a string of tokens that all have the same precedence (but different associativity).

I don't disagree with you at all that there are better ways to do this.  What I wanted to do first was a proof-of-concept, to see whether the bones in Pint and Pint-Pandas were ready for this sort of thing.  And I think they are.  My principle concern in modifying the tokenizer was out of respect for the fact that Pint does talk to itself using str() and __repr__, and as such we need to be prepared to handle whatever sorts of formats users may throw at us (including the N(S) notation and paranthesized ± as a unicode symbol).

Please keep the thoughts coming!

**Comment by hgrecco** [comment]
> On the choice of creating a tighter binding than **, consider `x ** 2+/-0 ** (1/2)'. If +/- binds the same as **, it will fail when there's a string of tokens that all have the same precedence (but different associativity).

I am not sure that I am proposing the same thing. Consider parsing `8 + / - 4`:
```python
TokenInfo(type=2 (NUMBER), string='8', start=(1, 0), end=(1, 1), line='8 + / - 4')
TokenInfo(type=54 (OP), string='+', start=(1, 2), end=(1, 3), line='8 + / - 4')
TokenInfo(type=54 (OP), string='/', start=(1, 4), end=(1, 5), line='8 + / - 4')
TokenInfo(type=54 (OP), string='-', start=(1, 6), end=(1, 7), line='8 + / - 4')
TokenInfo(type=2 (NUMBER), string='4', start=(1, 8), end=(1, 9), line='8 + / - 4')
```

which is transformed to:

```python
TokenInfo(type=2 (NUMBER), string='8', start=(1, 0), end=(1, 1), line='8 + / - 4')
TokenInfo(type=<NEW NUMBER> (OP), string='+ / -', start=(1, 2), end=(1, 7), line='8 + / - 4')
TokenInfo(type=2 (NUMBER), string='4', start=(1, 8), end=(1, 9), line='8 + / - 4')
```

which is then converted by `build_eval_tree` to:
```python
ufloat_fromstr("8 +/- 4")
```
This will automatically support every format supported [here](https://pythonhosted.org/uncertainties/user_guide.html#creating-numbers-with-uncertainties). I think we should limit to those (at least for the time being).

---
Consider now parsing `(8 + / - 4) m`:

```python
TokenInfo(type=54 (OP), string='(', start=(1, 0), end=(1, 1), line='(8 +/- 4) m')
TokenInfo(type=2 (NUMBER), string='8', start=(1, 1), end=(1, 2), line='(8 +/- 4) m')
TokenInfo(type=54 (OP), string='+', start=(1, 3), end=(1, 4), line='(8 +/- 4) m')
TokenInfo(type=54 (OP), string='/', start=(1, 4), end=(1, 5), line='(8 +/- 4) m')
TokenInfo(type=54 (OP), string='-', start=(1, 5), end=(1, 6), line='(8 +/- 4) m')
TokenInfo(type=2 (NUMBER), string='4', start=(1, 7), end=(1, 8), line='(8 +/- 4) m')
TokenInfo(type=54 (OP), string=')', start=(1, 8), end=(1, 9), line='(8 +/- 4) m')
TokenInfo(type=1 (NAME), string='m', start=(1, 10), end=(1, 11), line='(8 +/- 4) m')
```

which will then be converted
```python
TokenInfo(type=54 (OP), string='(', start=(1, 0), end=(1, 1), line='(8 +/- 4) m')
TokenInfo(type=2 (NUMBER), string='8', start=(1, 1), end=(1, 2), line='(8 +/- 4) m')
TokenInfo(type=<NEW NUMBER> (OP), string='+ / -', start=(1, 3), end=(1, 6), line='(8 +/- 4) m')
TokenInfo(type=2 (NUMBER), string='4', start=(1, 7), end=(1, 8), line='(8 +/- 4) m')
TokenInfo(type=54 (OP), string=')', start=(1, 8), end=(1, 9), line='(8 +/- 4) m')
TokenInfo(type=1 (NAME), string='m', start=(1, 10), end=(1, 11), line='(8 +/- 4) m')
```

which is then converted by `build_eval_tree` to:
```python
ufloat_fromstr("8 +/- 4") * ureg.meter
```
(I am 'paraphrasing' here) 


_Please note that that Pint supports more numerical types than uncertainties. We need to make this clear in the docs_

**Comment by MichaelTiemannOSC** [comment]
I have returned from a long trip and am ready to tackle this now.  I like the proposed approach and will give it a go.

**Comment by MichaelTiemannOSC** [comment]
The original pint tokenizer is `_plain_tokenizer` and the tokenizer that layers on top to handle uncertainties is `uncertainty_tokenizer`.  I've created a `tokenizer` variable so that users can switch away from the default this way:

```
from pint import pint_eval
pint_eval.tokenizer = pint_eval.uncertainty_tokenizer
```

I'm very open to changing names and/or mechanics of how the tokenizer should be selected.  I've not yet started to flesh out some better test cases for test_measurement.py.  I'd like to get feedback on these changes first.

File: pint/pint_eval.py [hunk 1]
```diff
@@ -9,14 +9,24 @@
...
+    HAS_UNCERTAINTIES = True
+except ImportError:
+    HAS_UNCERTAINTIES = False
+    ufloat = None
+
 from .errors import DefinitionSyntaxError
 
 # For controlling order of operations
 _OP_PRIORITY = {
+    "±": 4,
```
Comment (MichaelTiemannOSC) [inline pint/pint_eval.py:28]: This line can go away: we change ± to +/- to avoid gratuitous problems with ERRORTOKEN.

**Comment by MichaelTiemannOSC** [comment]
Any comments on this PR?  I think I made the changes that were requested.

**Comment by MichaelTiemannOSC** [comment]
I'm still waiting, and my project is still waiting to know if this is ready to enter final approach.  We'd really like to add uncertainties to our pint-based work, but are loathe to ask users to install a completely custom version of pint.  If there are other changes you would like to see, I would be happy to receive feedback!

File: pint/pint_eval.py [hunk 2]
```diff
@@ -45,7 +61,203 @@ def _power(left, right):
...
+            yield plus_minus_op
+            yield std_dev
+        else:
+            yield tokinfo
+
+
+tokenizer = _plain_tokenizer
+
 _BINARY_OPERATOR_MAP = {
+    "±": _ufloat,
```
Comment (MichaelTiemannOSC) [inline pint/pint_eval.py:259]: The line `"±": _ufloat,` can probably also go away, as above.

File: pint/pint_eval.py [hunk 3]
```diff
@@ -45,7 +61,203 @@ def _power(left, right):
...
+                return None
+        return self.buffer[n]
+
+
+def _plain_tokenizer(input_string):
+    for tokinfo in tokenize.tokenize(BytesIO(input_string.encode("utf-8")).readline):
+        if tokinfo.type != tokenlib.ENCODING:
+            yield tokinfo
+
+def uncertainty_tokenizer(input_string):
```
Comment (MichaelTiemannOSC) [inline pint/pint_eval.py:94]: Would `uncertainties_tokenizer` be a better name?

**Comment by MichaelTiemannOSC** [comment]
Alas there are some things that the merge did not correctly square.  If I had permission to do so, I'd convert this PR back to Draft status until they are ironed out.

**Comment by MichaelTiemannOSC** [comment]
Can somebody please review?

**Comment by jakeogh** [comment]
While trying to understand how to fix #1646 I pulled this branch and ran the tests:

Before parse-uncertainties:
```
platform linux -- Python 3.10.11, pytest-7.3.0, pluggy-1.0.0                                                                                                                                                                                                                     
rootdir: /home/cfg/_myapps/pint                                                                                                                                                                                                                                                  
plugins: typeguard-3.0.2, pkgcore-0.12.20, anyio-3.6.1                                                                                                                                                                                                                           
collected 2080 items / 2 skipped
<snip>
2 failed, 2024 passed, 13 skipped, 11 xfailed, 11 warnings, 32 errors
```

After parse-uncertainties:
```
platform linux -- Python 3.10.11, pytest-7.3.0, pluggy-1.0.0                                                                                                                                                                                                                     
rootdir: /home/cfg/_myapps/pint                                                                                                                                                                                                                                                  
plugins: typeguard-3.0.2, pkgcore-0.12.20, anyio-3.6.1                                                                                                                                                                                                                           
collected 2083 items / 2 skipped
11 failed, 2018 passed, 13 skipped, 11 xfailed, 13 warnings, 32 errors
```

pint-convert is unfortunately still broken (I rebased, so no surprise):
```
$ pint-convert 1ft meter
Traceback (most recent call last):
  File "/usr/lib/python-exec/python3.10/pint-convert", line 109, in <module>
    ureg._units["R_inf"].converter.scale = R_i
  File "<string>", line 4, in __setattr__
dataclasses.FrozenInstanceError: cannot assign to field 'scale'
```

**Comment by MichaelTiemannOSC** [comment]
@jakeogh I'm only just seeing your comment now.  No idea why I was not notified by GitHub.

**Comment by hgrecco** [comment]
I like this but changing the parser is always tricky. I would rather not merge it in this last version for Python 3.8. Flagging for the next

**Comment by MichaelTiemannOSC** [comment]
Results of running test-suite post-merge:

```
FAILED testsuite/test_compat.py::test_isnan[False] - TypeError: must be real number, not datetime.datetime
FAILED testsuite/test_compat.py::test_isnan[True] - TypeError: must be real number, not datetime.datetime
FAILED testsuite/test_compat.py::test_zero_or_nan[False] - TypeError: must be real number, not datetime.datetime
FAILED testsuite/test_compat.py::test_zero_or_nan[True] - TypeError: must be real number, not datetime.datetime
FAILED testsuite/test_issues.py::TestIssues::test_issue783 - ValueError: cannot call `vectorize` on size 0 inputs unless `otypes` is set
FAILED testsuite/test_quantity.py::TestTimedelta::test_add_sub - TypeError: must be real number, not datetime.datetime
FAILED testsuite/test_quantity.py::TestTimedelta::test_iadd_isub - TypeError: must be real number, not datetime.datetime
FAILED testsuite/test_unit.py::TestRegistry::test_parse_number - TypeError: can't convert complex to float
FAILED testsuite/test_unit.py::TestRegistryWithDefaultRegistry::test_parse_number - TypeError: can't convert complex to float
===================================================================== 9 failed, 2078 passed, 17 skipped, 11 xfailed, 12 warnings, 365 subtests passed in 99.78s (0:01:39) =====================================================================
(itr_env) iMac-Pro:pint michael$ 
```

**Comment by MichaelTiemannOSC** [comment]
I don't know why Babel is not installed in the CI/CD environment (which then causes failures in test_issues).  I have it installed in my environment, and test_issues passes.  I see other failures (which are skipped by the test suite).  I don't know how to align my world with the CI/CD world.

```
=========================================================================================================================== short test summary info ============================================================================================================================
FAILED pint/testsuite/test_compat.py::test_isnan[False] - TypeError: must be real number, not datetime.datetime
FAILED pint/testsuite/test_compat.py::test_isnan[True] - TypeError: must be real number, not datetime.datetime
FAILED pint/testsuite/test_compat.py::test_zero_or_nan[False] - TypeError: must be real number, not datetime.datetime
FAILED pint/testsuite/test_compat.py::test_zero_or_nan[True] - TypeError: must be real number, not datetime.datetime
FAILED pint/testsuite/test_issues.py::TestIssues::test_issue783 - ValueError: cannot call `vectorize` on size 0 inputs unless `otypes` is set
FAILED pint/testsuite/test_quantity.py::TestTimedelta::test_add_sub - TypeError: must be real number, not datetime.datetime
FAILED pint/testsuite/test_quantity.py::TestTimedelta::test_iadd_isub - TypeError: must be real number, not datetime.datetime
FAILED pint/testsuite/test_unit.py::TestRegistry::test_parse_number - TypeError: can't convert complex to float
FAILED pint/testsuite/test_unit.py::TestRegistryWithDefaultRegistry::test_parse_number - TypeError: can't convert complex to float
SKIPPED [1] pint/testsuite/test_compat_downcast.py:8: sparse is not available
SKIPPED [1] pint/testsuite/test_compat_upcast.py:6: xarray is not available
SKIPPED [1] pint/testsuite/test_dask.py:11: Dask is not available
SKIPPED [1] pint/testsuite/test_babel.py:9: Requires Babel not to be installed
SKIPPED [1] pint/testsuite/test_issues.py:168: Requires NumPy not to be installed.
SKIPPED [1] pint/testsuite/test_measurement.py:10: Requires Uncertainties not to be installed.
SKIPPED [6] pint/testsuite/test_numpy.py: unconditional skip
SKIPPED [2] pint/testsuite/test_quantity.py:372: Requires MIP
SKIPPED [2] pint/testsuite/test_quantity.py:410: Requires MIP
SKIPPED [2] pint/testsuite/test_quantity.py:425: Requires MIP
SKIPPED [2] pint/testsuite/test_quantity.py:684: Requires NumPy not to be installed.
XFAIL pint/testsuite/test_issues.py::TestIssues::test_issue25
XFAIL pint/testsuite/test_issues.py::TestIssues::test_issue37
XFAIL pint/testsuite/test_issues.py::TestIssues::test_issue39
XFAIL pint/testsuite/test_issues.py::TestIssues::test_issue86c
XFAIL pint/testsuite/test_log_units.py::test_compound_log_unit_multiply_definition
XFAIL pint/testsuite/test_log_units.py::test_compound_log_unit_quantity_definition
XFAIL pint/testsuite/test_log_units.py::test_dbm_db_addition
XFAIL pint/testsuite/test_log_units.py::test_frequency_octave_addition[100-2.0-400]
XFAIL pint/testsuite/test_log_units.py::test_frequency_octave_addition[50-1.0-100]
XFAIL pint/testsuite/test_log_units.py::test_frequency_octave_addition[200-0.0-200]
XFAIL pint/testsuite/test_numpy.py::TestNumpyMathematicalFunctions::test_exponentiation_array_exp_2
===================================================================================== 9 failed, 2081 passed, 20 skipped, 11 xfailed, 18 warnings, 365 subtests passed in 99.19s (0:01:39) ======================================================================================
(itr_env) iMac-Pro:pint michael$ pip list | grep -i babel
Babel                         2.12.1
```

**Comment by keewis** [comment]
you need to pin `babel` to `<2.9`, since `pint` is only compatible with `babel<=2.8`. Might be time to adapt to newer versions of `babel`, but as always somebody has to take the time to actually do it.

**Comment by MichaelTiemannOSC** [comment]
Shall we merge?

**Comment by hgrecco** [comment]
Yes but Any idea why the test suite is not running for this PR?

**Comment by MichaelTiemannOSC** [comment]
I don't know how the CI/CD works.  I certainly haven't tried to do anything to evade it.  Does anybody else know?

**Comment by keewis** [comment]
as far as I can tell, for PRs of new contributors (as far as the commit history is concerned) the maintainers have to manually confirm every time there's commits.

This is configurable (we have switched it off in xarray), but I think github's intention is to by default avoid running CI with potentially malicious changes before having a maintainer review it.

**Comment by keewis** [comment]
it is strange, though, that it doesn't tell you that. Maybe try closing / reopening the PR?

**Comment by MichaelTiemannOSC** [comment]
Closing PR to reopen it (attempting soft reboot of CI/CD).

**Comment by MichaelTiemannOSC** [comment]
Reopening PR to attempt soft reboot of CI/CD.

**Comment by MichaelTiemannOSC** [comment]
It looks like there's a problem with a readthedocs key on Pint's side.

**Comment by keewis** [comment]
no, it's just that the configuration (`.readthedocs.yaml`) needs an update. As far as I can tell, `system_packages: false` can be dropped without a change in behavior. This would be an easy fix if you want to send in a PR.

After that is merged, you also wouldn't have to wait on maintainer approval for CI to run on this PR.

**Comment by hgrecco** [comment]
Now it seems to be running. If all test pass, I will merge right away!

**Comment by hgrecco** [comment]
Test are failing. There seems to be an error due to missing babel dependency for a test.

**Comment by MichaelTiemannOSC** [comment]
I'm taking a wild guess that because we now have a test case that depends on babel, we need babel as part of `testbase`.  So I'm adding that there and saying a small prayer.  If somebody knows how to actually fix this, I'm all ears!

**Comment by MichaelTiemannOSC** [comment]
Not my day: lint failed for a reason I don't understand/cannot reproduce: reserveCache failed: Cache already exists. Scope: refs/heads/parse-uncertainties, Key: [xxx], Version: 25d34710bdd63bed70a22b8204c3e1d0942dc237e098600e3896b4334c6d784f

**Comment by hgrecco** [comment]
It is better that you use the decorator that is provided for this purpose:

https://github.com/hgrecco/pint/blob/2852f36db9aab916dbaaa68232ef45efccac3426/pint/testsuite/helpers.py#L125

Just import `requires_babel` and add `@requires_babel` to any test that requires babel.

**Comment by MichaelTiemannOSC** [comment]
Thanks, I'll fix that.  There are some new failures I need to fix, which I will:

```
FAILED pint/testsuite/test_compat.py::test_isnan[False] - TypeError: must be real number, not datetime.datetime
FAILED pint/testsuite/test_compat.py::test_isnan[True] - TypeError: must be real number, not datetime.datetime
FAILED pint/testsuite/test_compat.py::test_zero_or_nan[False] - TypeError: must be real number, not datetime.datetime
FAILED pint/testsuite/test_compat.py::test_zero_or_nan[True] - TypeError: must be real number, not datetime.datetime
```

**Comment by hgrecco** [comment]
Looks good! Shall we merge @MichaelTiemannOSC ?

**Comment by MichaelTiemannOSC** [comment]
Yes, thanks 🙏

**Comment by hgrecco** [comment]
Congratulations and thanks for this excelent work!!

**Pull request merged**